File: SimplePerceptron/simple_perceptron.cpp
#include <iostream>
#include <stdio.h>
#include <string>
#include <vector>
#include <cmath>

using namespace std;

/* 訓練データの読み込み */
int train_data_load(const char *filename, vector<float> &x1, vector<float> &x2, vector<int> &label)
{
	FILE *fp;
	int i=0;
	int temp_label;
	float temp_x1, temp_x2;

	if((fp = fopen(filename, "r")) == NULL){
		cerr<<"open file faild."<<endl;
		return -1;
	}

	while(fscanf(fp, "%f %f %d", &temp_x1, &temp_x2, &temp_label) != EOF)
	{
		x1.push_back(temp_x1);
		x2.push_back(temp_x2);
		label.push_back(temp_label);
		i++;
	}
	return i;
}

/* シグモイド関数 */
float sigmoid(float x)
{
	return 1.0f / (1.0f + exp(-x));
}

/* シグモイド関数の微分 */
float sigmoid_dot(float x)
{
	return sigmoid(x)*(1.0f - sigmoid(x));
}

/* 誤差関数 */
float error_func(vector<float> x1, vector<float> x2, vector<int> label, float *w, const int data_num)
{
	float L=0.0;

	for(int i=0;i<data_num;i++)
		L += 0.5*pow((label[i]-sigmoid(x1[i]*w[0]+x2[i]*w[1])), 2);
	
	return L;
}

/* 重みベクトルの学習 */
void learning(vector<float> x1, vector<float> x2, vector<int> label, float &w1, float &w2, float *w_old, const int data_num, const int alpha=0.1)
{
	float w_new[2] = {0.0f, 0.0f};

	for(int i=0;i<data_num; i++){
		float x[2] = {x1[i], x2[i]};
		float wx=0.0;
		wx  = x1[i]*w_old[0] + x2[i]*w_old[1];
		for(int j=0;j<2;j++)
			w_new[j] = w_old[j] - alpha*(label[i]-sigmoid(wx))*sigmoid_dot(wx)*x[j];
	}
	w1 = w_new[0];
	w2 = w_new[1];
}

int main(int argc, char *argv[])
{
	vector<float> x1, x2;
	vector<int> label;
	float w_old[2] = {0.5, 0.2};
	float w[2];
	const int ITERATION=10;
	
	/* 訓練データのロード */
	const int data_num = train_data_load("input.txt", x1, x2, label);

	/* 重みベクトルの学習 */
	for(int i=0;i<ITERATION;i++){
		learning(x1, x2, label, w[0], w[1], w_old, data_num);
		for(size_t i=0;i<2;i++)
			w_old[i] = w[i];
		cout<<w[0]<<" "<<w[1]<<endl; 

		float L=0.0f;
		L = error_func(x1, x2, label, w, data_num);
		//cout<<"error:"<<L<<endl;
	}

	return 0;
}

File: SimplePerceptron/simple_perceptron.py
import numpy as np
import matplotlib.pyplot as plt

EPS = 0.000001

class SimplePerceptron:
    def __init__(self, data_num, w1, w2, rate):
        self.data_num = data_num
        self.w1 = w1
        self.w2 = w2
        self.rate = rate

        self.wold = [0.0, 0.0]

        self.data = []
        
    def train_data_load(self, filename):
        self.data = np.loadtxt(filename)

    def sigmoid(self, x):
        return 1 / (1 + np.exp(-x))

    def sigmoid_dot(self, x):
        return self.sigmoid(x)*(1-self.sigmoid(x))

    def error_func(self):
        L = 0.0
        for i in range(self.data_num):
            fx = self.sigmoid(self.w1*self.data[i][0]+self.w2*self.data[i][1])
            L += 0.5*pow((self.data[i][2]-fx),2)
        return L

    def train(self):
        wx = 0.0
        w = [0,0]    
        for i in range(self.data_num):
            xi = [self.data[i][0], self.data[i][1]]
            wx = self.wold[0]*xi[0]+self.wold[1]*xi[1]
            w[0] = self.wold[0] + self.rate * (self.data[i][2]-self.sigmoid(wx))*self.sigmoid_dot(wx)*xi[0]
            w[1] = self.wold[1] + self.rate * (self.data[i][2]-self.sigmoid(wx))*self.sigmoid_dot(wx)*xi[1]
        self.w1 = w[0]
        self.w2 = w[1]
        self.wold = w

    def run(self, iteration):
        for i in range(iteration):
            self.train()
            print self.error_func()

    def plot(self):
        x1 = []
        y1 = []
        x2 = []
        y2 = []

        for i in range(self.data_num):
            if self.data[i][2] == 1:
                x1.append(self.data[i][0])
                y1.append(self.data[i][1])
            else:
                x2.append(self.data[i][0])
                y2.append(self.data[i][1])

        xfig = range(-2, 6)
        yfig = [ -(self.w2/self.w1)*xi for xi in xfig]
        plt.scatter(x1, y1, marker='o', color='r')
        plt.scatter(x2, y2, marker='o', color='b')
        plt.plot(xfig, yfig)
        plt.show()

if __name__ == '__main__':
    sp = SimplePerceptron(100, 0.5, 0.1, 0.1)
    
    #try:
    sp.train_data_load('input.txt')
    sp.run(1000)
    #except Exception, e:
    #    print "Faild."

    sp.plot()

